### ISSUE TITLE
Docker deployment of several instances

### ISSUE THREAD

**Issue by DaRealCapz** [VictoriaMetrics/VictoriaMetrics#5092]
### Is your question request related to a specific component?

vmagent

### Describe the question in detail

Trying to create a setup in a docker environment where I have 3 vmagents scraping metrics from 10 different node exporters and sending the metrics to a single victoria metrics instance. Is there a way to create a docker-compose.yml file where I dont have to repeat these lines of code?

![imagem](https://github.com/VictoriaMetrics/VictoriaMetrics/assets/71236784/4331a34e-d0eb-401d-8914-56c3a6274bbd)

Following the same logic I would have to repeat this 9 more times.

### Troubleshooting docs

- [X] General - https://docs.victoriametrics.com/Troubleshooting.html
- [X] vmagent - https://docs.victoriametrics.com/vmagent.html#troubleshooting
- [ ] vmalert - https://docs.victoriametrics.com/vmalert.html#troubleshooting

**Comment by zekker6** [comment]
Hello @DaRealCapz 

> Is there a way to create a docker-compose.yml file where I dont have to repeat these lines of code?
You can use [YAML anchors](https://medium.com/@kinghuang/docker-compose-anchors-aliases-extensions-a1e4105d70bd) to reduce amount of copy-pasted entries.

So that configuration will look like:
```
node_exporter: &node_exporter
  # full configuration


node_exporter2:
  <<: *node_exporter
  container_name: node_exporter2
```

> I have tried it but it appears as down on the discovered targets.

I think you've misunderstood what is `9101:9100` is doing here. This part is configuring [docker networking](https://docs.docker.com/network/) to forward container port `9100` to port `9101` on host machine. However, inside containers network the port remains the same(`9100`).
So in order to make it work you need to use `node-exporter2:9100` address.

> Also, how can I assert that for instance vmagent1 only scrapes from node exporter 1, 2 and 3 while vmagent 2 scrapes from node exporter 4, 5 and 6

The easiest option is to have multiple configuration files each for respective node-exporter instances.

**Comment by DaRealCapz** [comment]
Thank you for the tips, 

> The easiest option is to have multiple configuration files each for respective node-exporter instances.

For this case, do I follow the same logic as you explained for node exporter? I keep the same port for all the vmagents?

Additionally, I was a little confused with the prometheus.yml file, in the docker compose file we specify for the vmagent component the scrape configs using the prometheus.yml file:
![imagem](https://github.com/VictoriaMetrics/VictoriaMetrics/assets/71236784/5790a9aa-571c-48d6-8743-f1bf24868d5a)

I dont understand why inside this file we need to have the configurations for the self vmagent, vmalert, victoria metrics and victorialogs, shouldnt it only be the targets that the vmagent is scraping metrics from for example node exporters?

**Comment by zekker6** [comment]
> For this case, do I follow the same logic as you explained for node exporter? I keep the same port for all the vmagents?

Yes, you can keep the same port for metrics scraping configuration.

> I dont understand why inside this file we need to have the configurations for the self vmagent, vmalert, victoria metrics and victorialogs, shouldnt it only be the targets that the vmagent is scraping metrics from for example node exporters?

This is added in order to scrape metrics and [monitor](https://docs.victoriametrics.com/Single-server-VictoriaMetrics.html#monitoring) another VictoriaMetrics components by default. You can remove those targets if they are not required for your setup.

**Comment by DaRealCapz** [comment]
Regarding the node exporters, I am still having trouble deploying several at the same time. 
[docker-compose.txt](https://github.com/VictoriaMetrics/VictoriaMetrics/files/12819657/docker-compose.txt)
[prometheus.txt](https://github.com/VictoriaMetrics/VictoriaMetrics/files/12819659/prometheus.txt)
I am using these configurations yet everytime I try to run the docker compose file I get this message:
![imagem](https://github.com/VictoriaMetrics/VictoriaMetrics/assets/71236784/55094ca6-6784-4835-8449-7f5f42012c99)

What am I doing wrong?

**Comment by zekker6** [comment]
The promlem here is `ports` section for node exporter:
```
    ports:
      - 9100:9100
```

This section tries to bind host port 9100 to container port 9100, but since there are multiple containers trying to do the same bind it fails.

In your case vmagent will be able to scrape node-exporter without specifying `ports` in docker-compose, so you can remove this.

Also, I'd suggest you to get more familiar with docker networking setup to make it easier to create configurations like this. You can use [this](https://docs.docker.com/get-started/resources/) resource as a starting point for your research.

**Comment by zekker6** [comment]
You need to use different configuration files for vmagent. Each of files should only configure scraping of the required node exporters, e.g. config X - node-exporter1-3, Y - node-exporter3-6, Z - node-exporter7-9.

**Comment by DaRealCapz** [comment]
Thank you for your help! It is solve.
